### PR TITLE
Remove multiSigSign and multiSigVerify

### PR DESCRIPTION
**Pull request by hewison-chris**
Replacing special functions for threshold multisig with use of new
`sign` and `verify` overloads. This is to make the code easier to reason
about.

### CODE REVIEW

File: source/agora/crypto/Schnorr.d [hunk 1]
```diff
@@ -230,6 +230,25 @@ public Signature sign (T) (
      */
     // Compute the challenge and reduce the hash to a scalar
     Scalar c = hashFull(const(Message!T)(X, R, data));
+    return sign(x, R, r, c);
+}
+
+/// sign with predefined message hash c (used for multisig with threshold)
+public Signature sign (in Scalar x, in Point R, in Scalar r, in Scalar c)
```
Comment (linked0) [inline source/agora/crypto/Schnorr.d:204]: I think the last parameter `c` should have a logical name. Its meaning is confused with other parameters, `x` or `r`.

File: source/agora/crypto/Schnorr.d [hunk 2]
```diff
@@ -254,21 +273,28 @@ public bool verify (T) (in Point X, in Signature sig, scope const auto ref T dat
     nothrow @nogc @trusted
 {
     Sig s = Sig.fromBlob(sig);
+    // Compute the challenge and reduce the hash to a scalar
+    Scalar c = hashFull(const(Message!T)(X, s.R, data));
+    return verify(X, Sig.fromBlob(sig), c);
+}
+
+/// verify with predefined message hash (used in threshold multisig)
+public bool verify (in Point X, in Sig sig, in Scalar c)
```
Comment (linked0) [inline source/agora/crypto/Schnorr.d]: Ditto

**Review by linked0 — APPROVED** [review]
LGTM with some requests.

File: source/agora/crypto/Schnorr.d (hunk 1 above)
Comment (hewison-chris) [reply source/agora/crypto/Schnorr.d:204]: As c is defined in the comment above the declaration I would prefer to keep it as c. Otherwise the contents of the function get difficult to read. I changed the comment to use `c` rather than c to help make it clearer. Also in Schnorr discussions the hashed message is often referred to as `c`.

**Comment by AndrejMitrovic** [comment]
If they no longer return `Signature` then why do we still need that struct?

Perhaps @Geod24 originally used `Sig` to avoid accidentally modifying parts of the signature? I don't recall the reasoning. But in any case, having both types is a bit confusing.

**Comment by hewison-chris** [comment]
> If they no longer return `Signature` then why do we still need that struct?
> 
> Perhaps @Geod24 originally used `Sig` to avoid accidentally modifying parts of the signature? I don't recall the reasoning. But in any case, having both types is a bit confusing.

I'll attempt to remove Signature and just pass around Sig in Agora. Using toBlob() and Sig.fromBlob() as required.

**Comment by Geod24** [comment]
IIRC the need for two structs came from the fact we used two different signature scheme. Now that we're only using Schnorr, one of them ~can~ should go.

**Comment by hewison-chris** [comment]
> IIRC the need for two structs came from the fact we used two different signature scheme. Now that we're only using Schnorr, one of them ~can~ should go.

So shall I rename `Sig` to `Signature` or keep it as Sig to remind us it is Schnorr?

**Comment by hewison-chris** [comment]
> > IIRC the need for two structs came from the fact we used two different signature scheme. Now that we're only using Schnorr, one of them ~can~ should go.
> 
> So shall I rename `Sig` to `Signature` or keep it as Sig to remind us it is Schnorr?

Or maybe even `SigRs`?

**Comment by Geod24** [comment]
Yes to the rename to `Signature`.

File: source/agora/crypto/Schnorr.d [hunk 3]
```diff
@@ -101,7 +101,7 @@ nothrow @nogc @safe unittest
 
 public Point extractNonce (Signature sig) @safe @nogc nothrow pure
 {
-    return Sig.fromBlob(sig).R;
+    return sig.R;
 }
```
Comment (Geod24) [inline source/agora/crypto/Schnorr.d]: Maybe just remove this function then ? It doesn't seem to bring any value.

File: source/agora/crypto/Schnorr.d [hunk 4]
```diff
@@ -121,17 +121,55 @@ public Point extractNonce (Signature sig) @safe @nogc nothrow pure
...
+        auto s = Scalar("0x2222222222222222222222222222222222222222222222222222222222222222");
+        auto sig = Signature(R, s);
+        assert(sig == Signature("0x11111111111111111111111111111111111111111111111111111111111111112222222222222222222222222222222222222222222222222222222222222222"),
+            format!"sig should = %s%s"(sig.R, sig.s.toString(PrintMode.Clear)));
+    }
 
-    /// Converts this signature to a BitBlob matching `agora.common.Types`
-    public Signature toBlob () const pure nothrow @nogc @safe
+    /// Converts this signature to a BitBlob
+    public BitBlob!512 toBlob () const pure nothrow @nogc @safe
```
Comment (Geod24) [inline source/agora/crypto/Schnorr.d]: Do we need `toBlob` anymore ? I think it can be removed.

File: source/agora/crypto/Schnorr.d [hunk 5]
```diff
@@ -121,17 +121,55 @@ public Point extractNonce (Signature sig) @safe @nogc nothrow pure
...
+    }
+
+    /// construct from its string representation
+    public this (in char[] str)
+    {
+        auto offset = str.length == 130 ? 2 : 0;
+        assert(str[offset .. $].length == 128, format!"Signature hex string should be 128 characters (or 130 with 0x) not %s"(str[offset .. $].length));
+        this.R = Point(str[offset .. offset + 2 * Point.sizeof]);
+        this.s = Scalar(str[offset + 2 * Point.sizeof .. $]);
+    }
```
Comment (Geod24) [inline source/agora/crypto/Schnorr.d]: Is this used ? Same question for the other ctor.

File: source/agora/crypto/Schnorr.d [hunk 6]
```diff
@@ -121,17 +121,55 @@ public Point extractNonce (Signature sig) @safe @nogc nothrow pure
...
     /// Commitment
     public Point R;
     /// Proof
     public Scalar s;
 
-    static assert(Signature.sizeof == typeof(this).sizeof);
+    static assert(BitBlob!64.sizeof == typeof(this).sizeof);
+
+    /// construct from Point and Scalar
+    public this (in Point R, Scalar s) pure nothrow @nogc @safe
```
Comment (Geod24) [inline source/agora/crypto/Schnorr.d]: Why is only one of them `in` ?

File: source/agora/crypto/Schnorr.d [hunk 7]
```diff
@@ -199,10 +237,29 @@ public Signature sign (T) (
...
      */
     // Compute the challenge and reduce the hash to a scalar
-    Scalar c = hashFull(const(Message!T)(X, R, data));
+    const Scalar c = hashFull(const(Message!T)(X, R, data));
+    return sign(x, R, r, c);
+}
+
+/// sign with prepared message hash `c` (used for multisig with threshold)
+public Signature sign (in Scalar x, in Point R, in Scalar r, in Scalar c)
+    nothrow @nogc @trusted
```
Comment (Geod24) [inline source/agora/crypto/Schnorr.d]: Why `@trusted` ?

File: source/agora/crypto/Schnorr.d [hunk 8]
```diff
@@ -247,26 +310,26 @@ nothrow @nogc @safe unittest
 {
     Pair kp = Pair.fromScalar(Scalar(`0x074360d5eab8e888df07d862c4fc845ebd10b6a6c530919d66221219bba50216`));
     static immutable string message = "Bosagora:-)";
-    auto signature = sign(kp, message);
-    assert(verify(kp.V, signature, message));
+    auto sig = sign(kp, message);
+    assert(verify(kp.V, sig, message));
```
Comment (Geod24) [inline source/agora/crypto/Schnorr.d]: Was this necessary ? Same for the next 2 unittests.

**Review by Geod24 — COMMENTED** [review]
This is doing quite a few things all at once which makes it hard to follow. E.g. `Sig` => `Signature` could be done independently.

File: source/agora/crypto/Schnorr.d (hunk 3 above)
Comment (hewison-chris) [reply source/agora/crypto/Schnorr.d]: Removed

File: source/agora/crypto/Schnorr.d (hunk 5 above)
Comment (hewison-chris) [reply source/agora/crypto/Schnorr.d]: I removed this function although do have a static `fromString` to create Signatures from hex strings.

File: source/agora/crypto/Schnorr.d (hunk 6 above)
Comment (hewison-chris) [reply source/agora/crypto/Schnorr.d]: Updated

File: source/agora/crypto/Schnorr.d (hunk 7 above)
Comment (hewison-chris) [reply source/agora/crypto/Schnorr.d]: Updated this a a couple of other functions.

File: source/agora/crypto/Schnorr.d [hunk 9]
```diff
@@ -99,9 +99,9 @@ nothrow @nogc @safe unittest
 
 *******************************************************************************/
 
-public Point extractNonce (Signature sig) @safe @nogc nothrow pure
+public Point extractNonce (Sig sig) @safe @nogc nothrow pure
 {
```
Comment (hewison-chris) [inline source/agora/crypto/Schnorr.d]: This gets removed in a later commit.

File: source/agora/crypto/Schnorr.d (hunk 8 above)
Comment (hewison-chris) [reply source/agora/crypto/Schnorr.d]: Updated

File: source/agora/crypto/Schnorr.d [hunk 10]
```diff
@@ -121,28 +104,49 @@ public Point extractNonce (Signature sig) @safe @nogc nothrow pure
```
Comment (AndrejMitrovic) [inline source/agora/crypto/Schnorr.d]: The documentation is now outdated (it talks about binary blobs).

Comment (hewison-chris) [reply source/agora/crypto/Schnorr.d]: Updated.

File: source/agora/crypto/Schnorr.d [hunk 11]
```diff
@@ -130,6 +130,26 @@ public struct Sig
 
     static assert(Signature.sizeof == typeof(this).sizeof);
 
+    /// construct from Point and Scalar
+    public this (in Point R, in Scalar s) pure nothrow @nogc @safe
+    {
+        this.R = R;
+        this.s = s;
+    }
```
Comment (Geod24) [inline source/agora/crypto/Schnorr.d:117]: This gets generated automatically. But let's leave it as is for now.

File: source/agora/crypto/Schnorr.d [hunk 12]
```diff
@@ -130,6 +130,26 @@ public struct Sig
...
+
+    unittest
+    {
+        import std.format;
+
+        auto R = Point("0x921405afbfa97813293770efd55865c01055f39ad2a70f2b7a04ac043766a693");
+        assert(R.isValid()); // valid Point
+        auto s = Scalar("0x074360d5eab8e888df07d862c4fc845ebd10b6a6c530919d66221219bba50216");
+        assert(s.isValid()); // valid Scalar
+        auto sig = Sig(R, s);
```
Comment (Geod24) [inline source/agora/crypto/Schnorr.d]: Testing actor seems a bit overkill but it makes for a good documentation I guess :P

**Review by Geod24 — APPROVED** [review]
Love it!
